keps/sig-storage/3107-csi-nodeexpandsecret.md/3107-csi-nodeExpandSecret.md

```diff
@@ -0,0 +1,277 @@
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+- [Infrastructure Needed (Optional)](#infrastructure-needed-optional)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+## Summary
+
+This KEP proposes a way to add NodeExpandSecret to the CSI persistent
+volume source and thus enabling the csi client to send it out as part of
+the nodeExpandVolume request to the csi drivers for making use of it
+in the various Node Operations.
+
+## Motivation
+
+### Goals
+
+- Introduce `secretRef` in CSI Persistent Volume Source.
+- Allow CSI driver to get/refer `secretRef` sent
+  from kubelet as part of `NodeExpandVolume` operation.
+- To support per-PVC secrets for volume resizing, similar to CSI attach and
+  detach - this proposal expands `CSIPersistentVolumeSource` object to
+  contain `NodeExpandSecretRef`.
+
+### Non-Goals
+
+- Other CSI calls e.g. `NodeStageVolume` will not have the secretRef
+  in the request, this is limited to `NodeExpandVolume` operation.
+
+## Proposal
+
+Currently, the CSI drivers dont have a method to make use of secretRef
+at time of Node operation (ex: nodeExpansion) as the subjected csi request does
+not carry a secret or credentials in the request. Even-though
+we have implemented similar mechanism for Controller side operations, ie secretRef
+field available in the csi PV source and making use of it while 
+controllerExpand request has been sent to the CSI driver,  similar field 
+is missing in the nodeExpansion request.
+
+### User stories
+
+#### story 1
+- At times, we need to check the actual size of the backend volume/image before
+  proceeding on FS resize to avoid false positive returns on fs resize operation.
+
+#### story 2
+- Encrypted device with LUKs, which need the passphrase in order to resize
+  the device on the node.
+
+#### story 3
+- For various validations at time of node expansion the CSI driver has to be connected
+  to the backend storage cluster, if the secretRef is part of the nodeExpansion request
+  the CSI driver can make use of the same and connect to the storage cluster 
+  to perform the cluster operations.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+## Design Details
+
+```go
+- pkg/apis/core/types.go
+..
+type CSIPersistentVolumeSource struct {
+    .....
+    // nodeExpandSecretRef is a reference to secret object containing sensitive
+    // information to pass to the CSI driver to complete CSI node expansion
+    NodeExpandSecretRef *SecretReference
+}
+```
+The above field NodeExpandSecretRef is optional:
+
+To enable, NodeExpandSecretRef a new feature gate (NodeExpandSecret) has to be
```

Comment:
I would probably call it "CSINodeExpandSecret"

Reply:
me too :)